bnd-maven-plugin-parent/bnd-maven-plugin/src/main/java/org/bndtools/maven/BndMavenPlugin.java

```diff
@@ -0,0 +1,186 @@
+	private File classesDir;
+	
+	@Parameter(defaultValue = "${project.build.outputDirectory}/META-INF/MANIFEST.MF", readonly = true)
+	private File manifestPath;
+	
+	@Parameter(defaultValue = "${project}", required = true, readonly = true)
+	private MavenProject project;
+	
+	public void execute() throws MojoExecutionException {
+		Log log = getLog();
+
+		File bndFile = new File(project.getBasedir(), Project.BNDFILE);
+
+		Builder builder = new Builder();
+		builder.setTrace(log.isDebugEnabled());
+		try {
+			if (bndFile.isFile())
+				builder.setProperties(bndFile, project.getBasedir());
+			else
+				builder.setBase(project.getBasedir());
+			
+			// Reject sub-bundle projects
+			Collection<? extends Builder> subs = builder.getSubBuilders();
+			if (subs.size() != 1)
+				throw new MojoExecutionException("Sub-bundles not permitted in a maven build");
+			File builderFile = builder.getPropertiesFile();
+			if (builderFile != null && !bndFile.equals(builderFile))
+				throw new MojoExecutionException("Sub-bundles not permitted in a maven build");
+
+			// Set bnd classpath
+			List<File> classpath = new LinkedList<File>();
+			@SuppressWarnings("unchecked")
+			Set<Artifact> artifacts = project.getArtifacts();
+			for (Artifact artifact : artifacts) {
+				File artifactFile = artifact.getFile();
+				if (artifactFile != null)
+					classpath.add(artifactFile);
+			}
+			classpath.add(classesDir);
+			builder.setClasspath(classpath.toArray(new File[classpath.size()]));
+
+			// Include local project packages automatically
+			String includes = builder.getProperty(Constants.INCLUDERESOURCE);
+			StringBuilder newIncludes = new StringBuilder().append('"').append(classesDir.getPath()).append('"');
+			if (includes == null || includes.trim().length() == 0)
+				includes = newIncludes.toString();
+			else
+				includes = newIncludes.append(',').append(includes).toString();
+			builder.setProperty(Constants.INCLUDERESOURCE, includes);
+
+			// Set Bundle-Version
+			MavenVersion mvnVersion = new MavenVersion(project.getVersion());
+			builder.setProperty(Constants.BUNDLE_VERSION, mvnVersion.toBndVersion());
+
+			// Build bnd Jar (in memory)
+			Jar bndJar = builder.build();
+			checkErrors(builder);
+
+			// Output manifest to <classes>/META-INF/MANIFEST.MF
+			Files.createDirectories(manifestPath.toPath().getParent());
+			FileOutputStream manifestOut = new FileOutputStream(manifestPath);
+			try {
+				bndJar.writeManifest(manifestOut);
+			} finally {
+				manifestOut.close();
+			}
+
+			// Expand Jar into target/classes
+			expandJar(bndJar, classesDir);
+			
+			// Check errors again since something might have gone wrong
+			// while writing manifest or JAR content
+			checkErrors(builder);
+
+		} catch (Exception e) {
+			throw new MojoExecutionException("bnd error", e);
+		} finally {
+			builder.close();
+		}
+	}
+
+	private void checkErrors(Builder builder) throws MojoExecutionException {
```

Comment:
That's probably fine. Checking after build() but before writing just saves some time before you fail the build.